From 7f09bcb1e795e58a6f1fe525a20526e5a2430bc6 Mon Sep 17 00:00:00 2001
From: Ryan Zezeski <rzezeski@gmail.com>
Date: Tue, 12 Feb 2013 11:54:21 -0500
Subject: [PATCH] Disable stale check and nagle

* Apache HTTP's stale conn check causes additionall latency and
  is not recommended for high-throughput/low-latency scenarios.

* Disabling the stale check requires adding a periodic background
  task which clears idles connections.  This prevents the client
  from pulling a conn closed by the server which causes an IOException.

* Disable nagle as it's meant for protocols that use many small messages.
---
 .../solr/handler/component/HttpShardHandlerFactory.java     |  3 +++
 .../org/apache/solr/client/solrj/impl/HttpClientUtil.java   | 13 +++++++++++++
 2 files changed, 16 insertions(+)

diff --git a/solr/core/src/java/org/apache/solr/handler/component/HttpShardHandlerFactory.java b/solr/core/src/java/org/apache/solr/handler/component/HttpShardHandlerFactory.java
index d55e01f..cc59731 100644
--- a/solr/core/src/java/org/apache/solr/handler/component/HttpShardHandlerFactory.java
+++ b/solr/core/src/java/org/apache/solr/handler/component/HttpShardHandlerFactory.java
@@ -29,6 +29,7 @@ import java.util.concurrent.SynchronousQueue;
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
 
+import org.apache.http.params.HttpConnectionParams;
 import org.apache.http.client.HttpClient;
 import org.apache.solr.client.solrj.SolrServerException;
 import org.apache.solr.client.solrj.impl.HttpClientUtil;
@@ -150,6 +151,8 @@ public class HttpShardHandlerFactory extends ShardHandlerFactory implements org.
     clientParams.set(HttpClientUtil.PROP_CONNECTION_TIMEOUT, connectionTimeout);
     clientParams.set(HttpClientUtil.PROP_USE_RETRY, false);
     this.defaultClient = HttpClientUtil.createClient(clientParams);
+    this.defaultClient.getParams().setParameter(HttpConnectionParams.STALE_CONNECTION_CHECK, false);
+    this.defaultClient.getParams().setParameter(HttpConnectionParams.TCP_NODELAY, true);
     this.loadbalancer = createLoadbalancer(defaultClient);
   }
 
diff --git a/solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpClientUtil.java b/solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpClientUtil.java
index 3887302..4649138 100644
--- a/solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpClientUtil.java
+++ b/solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpClientUtil.java
@@ -18,6 +18,9 @@ package org.apache.solr.client.solrj.impl;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
 import java.util.zip.GZIPInputStream;
 import java.util.zip.InflaterInputStream;
 
@@ -102,6 +105,16 @@ public class HttpClientUtil {
     final ModifiableSolrParams config = new ModifiableSolrParams(params);
     logger.info("Creating new http client, config:" + config);
     final DefaultHttpClient httpClient = new SystemDefaultHttpClient();
+    final org.apache.http.conn.ClientConnectionManager mgr = httpClient.getConnectionManager();
+
+    // NOTE: The sweeper task is assuming hard-coded Jetty max-idle of 50s.
+    final Runnable sweeper = new Runnable() {
+            public void run() {
+                mgr.closeIdleConnections(40, TimeUnit.SECONDS);
+            }
+        };
+    final ScheduledExecutorService stp = Executors.newScheduledThreadPool(1);
+    stp.scheduleWithFixedDelay(sweeper, 5, 5, TimeUnit.SECONDS);
     configureClient(httpClient, config);
     return httpClient;
   }
-- 
1.8.1.2

